publickey.go

```diff
@@ -133,6 +133,34 @@ func (k *PublicKey) Hex(compressed bool) string {
 	return hex.EncodeToString(k.Bytes(compressed))
 }
 
+// DecapsulateKEM decapsulates key by using Key Encapsulation Mechanism and returns symmetric key;
+// can be safely used as encryption key
+func (k *PublicKey) DecapsulateKEM(priv *PrivateKey) ([]byte, error) {
+	if priv == nil {
+		return nil, errors.New("public key is empty")
+	}
+
+	var secret bytes.Buffer
+	secret.Write(k.Bytes(false))
+
+	sx, sy := priv.Curve.ScalarMult(k.X, k.Y, priv.D.Bytes())
+	secret.Write([]byte{0x04})
+
+	// Sometimes shared secret coordinates are less than 32 bytes; Big Endian
+	l := len(priv.Curve.Params().P.Bytes())
```

Comment:
better extract this part like a `padZeros` function

Reply:
@kigawas do you mean bit-shifting or what?

Reply:
I mean there are some duplicate codes of padding zeros like:
```golang
	for i := 0; i < l-len(sx.Bytes()); i++ {
		secret.Write([]byte{0x00})
	}
	secret.Write(sx.Bytes())
```

Reply:
Oh, got it. But in case of Go I should also check errors :D So by extracting it into own function I will add checks. But thanks for pointing, I will do some refactor.